src/js/chromecast/ChromecastSessionManager.js

```diff
@@ -0,0 +1,122 @@
+            this.player.trigger('chromecastDisconnected');
+            this._reloadTech();
+         }
+      }.bind(this));
+
+      // We must create these objects here, after setting the options on the cast context
+      // (see the `configureCastContext` function in enableChromecast.js) but before
+      // requesting a session or loading media. For an undocumented (and thus unknown)
+      // reason, RemotePlayer and RemotePlayerController instances created later (for
+      // example, right after successfully creating and connecting to a session, in the
+      // `openCastMenu` function) will not stay in sync with media items that are loaded
+      // later.
+      //
+      // For example, the first item that you cast will work as expected: events on
+      // RemotePlayerController will fire and the state (currentTime, duration, etc) of
+      // the RemotePlayer instance will update as the media item plays. However, if a new
+      // media item is loaded via a `loadMedia` request, the media item will play, but the
+      // remotePlayer will be in a "media unloaded" state where the duration is 0, the
+      // currentTime does not update, and no change events are fired (except, strangely,
+      // displayStatus updates).
+      this.remotePlayer = new cast.framework.RemotePlayer();
+      this.remotePlayerController = new cast.framework.RemotePlayerController(this.remotePlayer);
+   },
+
+   openCastMenu: function() {
+      var onSessionSuccess;
+
+      if (!this.player.currentSource()) {
+         // Do not cast if there is no media item loaded in the player
+         return;
+      }
+      onSessionSuccess = function() {
+         hasConnected = true;
+         this.player.trigger('chromecastConnected');
+         this._reloadTech();
+      }.bind(this);
+
+      // It is the `requestSession` function call that actually causes the cast menu to
+      // open.
+      // The second parameter to `.then` is an error handler. We use _.noop here
+      // because we handle errors in the ChromecastTech class and we do not want an
+      // error to bubble up to the console. This error handler is also triggered when
+      // the user closes out of the chromecast selector pop-up without choosing a
+      // casting destination.
+      this.getCastContext().requestSession()
+         .then(onSessionSuccess, _.noop);
+   },
+
+   _reloadTech: function() {
+      var player = this.player,
+          currentTime = player.currentTime(),
+          wasPaused = player.paused();
+
+      // Reload the current source to re-lookup and use the currently available Tech. The
+      // chromecast Tech gets used if `ChromecastSessionManager.isChromecastConnected` is
+      // true (effectively, if a chromecast session is currently in progress), otherwise
+      // Video.js continues to search through the Tech list for other eligible Tech to
+      // use, such as the HTML5 player.
+      player.src(_.extend({}, player.currentSource(), {
```

Comment:
Remember, in Video.js there is `this.cache_.source` and `this.cache_.src`.

- [`player.src()`](https://github.com/videojs/video.js/blob/998c0042cbca62b743df452bb8ffb5e659599589/src/js/player.js#L2267)

   ```js
   return this.cache_.src;
   ```

- [`player.currentSource()`](https://github.com/videojs/video.js/blob/998c0042cbca62b743df452bb8ffb5e659599589/src/js/player.js#L2417)

   ```js
   return this.cache_.source || {};
   ```

While `this.cache_.src` is [updated by the middleware](https://github.com/videojs/video.js/blob/998c0042cbca62b743df452bb8ffb5e659599589/src/js/player.js#L2315), `this.cache_.source` is not. `this.cache_.source` is [only set here](https://github.com/videojs/video.js/blob/998c0042cbca62b743df452bb8ffb5e659599589/src/js/player.js#L2287-L2288).

To demonstrate the issue, notice the following. The code adds another source and the middleware selected the last source from the given list of sources.

```diff
diff --git a/docs/demo/index.html b/docs/demo/index.html
index 207027c..c9641af 100644
--- a/docs/demo/index.html
+++ b/docs/demo/index.html
@@ -13,6 +13,7 @@
    <h1>Demo of <code>silvermine-videojs-chromecast</code></h1>

    <video id="video_1" class="video-js vjs-default-skin" controls preload="auto" data-setup='{ "fluid": "true" }'>
+      <source src="http://www.caminandes.com/download/01_llama_drama_1080p.mp4" type="video/mp4">
       <source src="http://www.caminandes.com/download/03_caminandes_llamigos_1080p.mp4" type="video/mp4">
    </video>

@@ -23,6 +24,23 @@
          techOrder: [ 'chromecast', 'html5' ],
       };

+      videojs.use('*', function(player) {
+         return {
+
+            setSource: function(playerSelectedSource, next) {
+               var sources = player.currentSources(),
+                   newSource = sources[sources.length - 1];
+
+               console.log('middleware: all sources', sources);
+               console.log('middleware: playerSelectedSource', playerSelectedSource.src);
+               console.log('middleware: middleware selected source', newSource.src);
+
+               next(null, newSource);
+            },
+
+         };
+      });
+
       videojs('video_1', options, function() {
          var player = this;

diff --git a/src/js/chromecast/ChromecastSessionManager.js b/src/js/chromecast/ChromecastSessionManager.js
index 8713d20..324e0cd 100644
--- a/src/js/chromecast/ChromecastSessionManager.js
+++ b/src/js/chromecast/ChromecastSessionManager.js
@@ -79,6 +79,8 @@ ChromecastSessionManager = Class.extend({
       // true (effectively, if a chromecast session is currently in progress), otherwise
       // Video.js continues to search through the Tech list for other eligible Tech to
       // use, such as the HTML5 player.
+      console.log('_reloadTech: currentSource.src', player.currentSource().src);
+      console.log('_reloadTech: src', player.src());
       player.src(_.extend({}, player.currentSource(), {
          duration: player.duration(),
          poster: player.poster(),
```

This is the outcome:

<img width="1094" alt="chromecast-middleware-source" src="https://user-images.githubusercontent.com/8900251/29327099-fecc3b0a-81bb-11e7-8092-cc77c9a3ec41.png">

While the correct video, `03_caminandes_llamigos_1080p.mp4`, plays for the user in the player. Once the chromecast plugin is triggered, the video source changes to `01_llama_drama_1080p.mp4`.